### PR TITLE
DBLINK changes

### PR DESCRIPTION
**Pull request by oualib**
- Option -ext is too much ambiguous and is removed.
 - Allowing multiple external connection: one for each symbol.
 - bug correction
 - new param for external vdf, pick up the right symbol

### CODE REVIEW

File: verticapy/toolbox.py [hunk 1]
```diff
@@ -1139,22 +1159,25 @@ def replace_external_queries_in_query(query: str):
...
-            external_query_tmp = external_query
-            subquery_flag = False
-        else:
-            external_query_tmp = f"SELECT * FROM {external_query}"
-            subquery_flag = True
-        query_dblink_template = get_dblink_fun(external_query_tmp)
-        if subquery_flag:
-            if " " in external_query.strip():
-                alias = f"VERTICAPY_EXTERNAL_TABLE_{idx}"
+    for s in verticapy.options["external_connection"]:
```
Comment (mail4umar) [inline verticapy/toolbox.py:1162]: not sure about this. 

don't you want to get the symbol inside the connection e.g. vertica.options["external_connection"["symbol"]

File: verticapy/toolbox.py [hunk 2]
```diff
@@ -1139,22 +1159,25 @@ def replace_external_queries_in_query(query: str):
...
-            query_dblink_template = f"({query_dblink_template}) AS {alias}"
-        query = query.replace(f"$$${external_query}$$$", query_dblink_template)
+                external_query_tmp = f"SELECT * FROM {external_query}"
+                subquery_flag = True
+            query_dblink_template = get_dblink_fun(external_query_tmp, symbol=s)
+            if subquery_flag:
+                if " " in external_query.strip():
+                    alias = f"VERTICAPY_EXTERNAL_TABLE_{idx}"
+                else:
+                    alias = '"' + external_query.strip().replace('"', '""') + '"'
```
Comment (mail4umar) [inline verticapy/toolbox.py:1176]: not sure when we will need this....

how can we have have " in the name of an external table?

File: verticapy/vdataframe.py [hunk 3]
```diff
@@ -125,6 +125,15 @@ class vDataFrame:
     A boolean to indicate whether it is an external table. If set to True, a
     Connection Identifier Database must be defined.
     See the connect.set_external_connection function for more information.
+symbol: str, optional
```
Comment (mail4umar) [inline verticapy/vdataframe.py:128]: we can also add here the exhaustive list of allowed symbols

File: verticapy/vdataframe.py [hunk 4]
```diff
@@ -220,11 +233,15 @@ def __init__(
...
                 ("schema", schema, [str]),
                 ("external", external, [bool]),
+                ("sql_push_ext", sql_push_ext, [bool]),
                 ("empty", empty, [bool]),
             ]
         )
 
         if external:
+            assert is_special_symbol(symbol), ParameterError(
+                "Parameter 'symbol' must be a special char. Example: $, €, ..."
```
Comment (mail4umar) [inline verticapy/vdataframe.py:243]: exhaustive list here

File: verticapy/vdataframe.py [hunk 5]
```diff
@@ -236,9 +253,17 @@ def __init__(
                     relation = str(input_relation)
                 cols = ", ".join(usecols) if usecols else "*"
                 query = f"SELECT {cols} FROM {input_relation}"
+
             else:
                 query = sql
-            sql = get_dblink_fun(query)
+
+            if symbol in verticapy.options["external_connection"]:
```
Comment (mail4umar) [inline verticapy/vdataframe.py:260]: i thought we have symbol inside the external connection.

**Review by mail4umar — APPROVED** [review]

**Review by tconsigny — APPROVED** [review]

File: verticapy/connect.py [hunk 6]
```diff
@@ -491,17 +491,28 @@ def set_external_connection(cid: str, rowset: int = 500):
     Connection Identifier Database.
 rowset: int, optional
     Number of rows retrieved from the remote database during each 
-    SQLFetch() cycle. 
+    SQLFetch() cycle.
+symbol: str, optional
+    Can be any special character except the SQL operators and parenthesis.
```
[the remark this answers is not in the document]
Comment (oualib) [reply verticapy/connect.py]: Unfortunately, I can not list them. Do you see a Pythonic way to list them?

File: verticapy/sql.py [hunk 7]
```diff
@@ -176,19 +166,22 @@ def sql(line, cell="", local_ns=None):
...
-            )
-            queries = get_dblink_fun(queries)
-
         # Cleaning the Query
         queries = clean_query(queries)
         queries = replace_vars_in_query(queries, locals()["local_ns"])
         queries = replace_external_queries_in_query(queries)
 
+        # Looking at very specific external queries symbols
+        for s in ["$", "€", "£"]:
```
[the remark this answers is not in the document]
Comment (oualib) [reply verticapy/sql.py:175]: same as up

File: verticapy/toolbox.py (hunk 1 above)
Comment (oualib) [reply verticapy/toolbox.py:1162]: verticapy.options["external_connection"] is a dict of symbols

**Review by mail4umar — APPROVED** [review]

**Pull request merged**